c10/core/SingletonSymNodeImpl.h

```diff
@@ -0,0 +1,121 @@
+#include <c10/core/ConstantSymNodeImpl.h>
+#include <c10/core/SymBool.h>
+#include <c10/core/SymNodeImpl.h>
+#include <iostream>
+
+namespace c10 {
+
+// An object that only defines the equality operator.
+class C10_API SingletonSymNodeImpl : public SymNodeImpl {
+ public:
+  SingletonSymNodeImpl(int64_t val) : val_(val) {}
+
+  bool bool_() override {
+    return false;
+  }
+
+  bool is_int() override {
+    return true;
+  }
+
+  bool is_float() override {
+    return false;
+  }
+
+  bool is_bool() override {
+    return false;
+  }
+
+  bool has_hint() override {
+    return true;
+  }
+
+  c10::SymNode wrap_int(int64_t num) override {
+    return SymNode(c10::make_intrusive<ConstantSymNodeImpl<int64_t>>(num));
+  };
+
+  int64_t guard_int(const char* file, int64_t line) override {
+    // TODO: when is this used?
+    TORCH_CHECK(false);
+  }
+
+  double guard_float(const char* file, int64_t line) override {
+    TORCH_CHECK(false, "not a float");
+  }
+
+  bool guard_bool(const char* file, int64_t line) override {
+    TORCH_CHECK(false, "not a bool");
+  }
+
+  int64_t int_() override {
+    // TODO: when is this used?
+    TORCH_CHECK(false);
+  }
+
+  std::string str() override {
+    return "j" + std::to_string(val_);
+  }
+
+  c10::SymNode eq(const c10::SymNode& other) override {
+    c10::optional<int64_t> c = other->singleton_int();
+    TORCH_CHECK(
+        c,
+        "SingletonSymNode can only be compared with SingletonSymNode, but got ",
+        other->str());
```

Comment:
It's not obvious to me you should error here, another reasonable answer would be to return false

Reply:
That's a good point. I haven't thought about this choice deeply, though maybe its ok to potentially relax later as a follow up?